src/observerClass.js

```diff
@@ -0,0 +1,161 @@
+    useStaticRenderingLite(useStaticRendering)
+}
+
+export function makeClassComponentObserver(componentClass) {
+    const target = componentClass.prototype
+    if (target.componentWillReact)
+        throw new Error("The componentWillReact life-cycle event is no longer supported")
+    if (componentClass.__proto__ !== PureComponent) {
+        if (!target.shouldComponentUpdate) target.shouldComponentUpdate = observerSCU
+        else if (target.shouldComponentUpdate !== observerSCU)
+            // n.b. unequal check, instead of existence check, as @observer might be on superclass as well
+            throw new Error(
+                "It is not allowed to use shouldComponentUpdate in observer based components."
+            )
+    }
+
+    // this.props and this.state are made observable, just to make sure @computed fields that
+    // are defined inside the component, and which rely on state or props, re-compute if state or props change
+    // (otherwise the computed wouldn't update and become stale on props change, since props are not observable)
+    // However, this solution is not without it's own problems: https://github.com/mobxjs/mobx-react/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3Aobservable-props-or-not+
+    makeObservableProp(target, "props")
+    makeObservableProp(target, "state")
+
+    const baseRender = target.render
+    target.render = function() {
+        return makeComponentReactive.call(this, baseRender)
+    }
+    patch(target, "componentWillUnmount", function() {
+        if (isUsingStaticRendering() === true) return
+        this.render[mobxAdminProperty] && this.render[mobxAdminProperty].dispose()
+        this[mobxIsUnmounted] = true
+    })
+    return componentClass
+}
+
+function makeComponentReactive(render) {
+    if (isUsingStaticRendering() === true) return render.call(this)
+
+    function reactiveRender() {
+        isRenderingPending = false
+        let exception = undefined
+        let rendering = undefined
+        reaction.track(() => {
+            try {
+                rendering = _allowStateChanges(false, baseRender)
+            } catch (e) {
+                exception = e
+            }
+        })
+        if (exception) {
+            throw exception
+        }
+        return rendering
+    }
+
+    // Generate friendly name for debugging
+    const initialName =
+        this.displayName ||
+        this.name ||
+        (this.constructor && (this.constructor.displayName || this.constructor.name)) ||
+        "<component>"
+    /**
+     * If props are shallowly modified, react will render anyway,
+     * so atom.reportChanged() should not result in yet another re-render
+     */
+    setHiddenProp(this, skipRenderKey, false)
+    /**
+     * forceUpdate will re-assign this.props. We don't want that to cause a loop,
+     * so detect these changes
+     */
+    setHiddenProp(this, isForcingUpdateKey, false)
+
+    // wire up reactive render
+    const baseRender = render.bind(this)
```

Comment:
Perhaps move this line above `reactiveRender` declaration? It's fairly confusing to follow this code with all that hoisting. I would even move `const reaction` on top for a natural flow.

Reply:
yeah it is only the `disposeOnUnmount` decorator support that introduces the actual complexity, not the class support itself (which is pretty simple, and doesn't need the ugly component-might-not-have-been-committed hack :))

Reply:
> doesn't need the ugly component-might-not-have-been-committed hack

Are you sure? I think the reaction would have to be created in `componentDidMount` (similary to inside `useEffect`).
Check https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#adding-event-listeners-or-subscriptions:
Specifically:
> async rendering (where rendering might be interrupted before it completes, causing componentWillUnmount not to be called).